File: advent_of_code/2022/02/day02.go
package main

import (
	"fmt"
	"os"
	"strings"
)

func part1(input string) int {
	win_score := 6
	draw_score := 3
	lose_score := 0
	rock_score := 1
	paper_score := 2
	scissor_score := 3

	sum := 0
	for _, line := range strings.Split(input, "\n") {
		line = strings.TrimSpace(line)
		switch line {
		case "A X":
			sum += draw_score + rock_score
		case "A Y":
			sum += win_score + paper_score
		case "A Z":
			sum += lose_score + scissor_score
		case "B X":
			sum += lose_score + rock_score
		case "B Y":
			sum += draw_score + paper_score
		case "B Z":
			sum += win_score + scissor_score
		case "C X":
			sum += win_score + rock_score
		case "C Y":
			sum += lose_score + paper_score
		case "C Z":
			sum += draw_score + scissor_score
		}
	}

	return sum
}

func part2(input string) int {
	win_score := 6
	draw_score := 3
	lose_score := 0
	rock_score := 1
	paper_score := 2
	scissor_score := 3

	sum := 0

	for _, line := range strings.Split(input, "\n") {
		line = strings.TrimSpace(line)
		switch line {
		case "A X":
			sum += lose_score + scissor_score
		case "A Y":
			sum += draw_score + rock_score
		case "A Z":
			sum += win_score + paper_score
		case "B X":
			sum += lose_score + rock_score
		case "B Y":
			sum += draw_score + paper_score
		case "B Z":
			sum += win_score + scissor_score
		case "C X":
			sum += lose_score + paper_score
		case "C Y":
			sum += draw_score + scissor_score
		case "C Z":
			sum += win_score + rock_score
		}
	}

	return sum
}

func main() {
	filename := "input.txt"
	const DEBUG = false
	var input string
	if DEBUG {
		input = `A Y
		B X
		C Z`
	} else {
		s, err := os.ReadFile(filename)
		if err != nil {
			panic(err)
		}
		input = string(s)
	}

	fmt.Printf("2022 Day 2, part 1: %v\n", part1(input))
	fmt.Printf("2022 Day 2, part 2: %v\n", part2(input))
}

File: advent_of_code/rust/src/aoc2015/day02.rs
type Gift = (u32, u32, u32);

#[allow(dead_code)]
fn parse_input(input: &str) -> Vec<Gift> {
  input.lines()
  .map(|line| line.trim())
  .map(|line| {
    let v = line.split("x").collect::<Vec<&str>>();
    match v[..] {
      [l, w, h] => (
        l.parse::<u32>().unwrap(),
        w.parse::<u32>().unwrap(),
        h.parse::<u32>().unwrap()
      ),
      _ => panic!("illegal pattern: {line}"),
    }
  }).collect()
}

#[allow(dead_code)]
pub fn part1(input: &Vec<Gift>) -> u32 {
  input.iter().map(|(l, w, h)| {
    let sides = vec![l * w, l * h, w * h];
    let smallest_side = sides.iter().min().unwrap();
    sides.iter().map(|v| *v * 2).sum::<u32>() + smallest_side
  }).sum()
}

#[allow(dead_code)]
pub fn part2(input: &Vec<Gift>) -> u32 {
  input.iter().map(|(l, w, h)| -> u32 {
    let mut sides = vec![*l, *w, *h];
    sides.sort();
    2 * sides.iter().take(2).sum::<u32>() + sides.iter().product::<u32>()
  }).sum()
}

#[allow(dead_code)]
pub fn main() {
  let input = include_str!("../../input/aoc2015/day02.input.txt");
  let gifts = parse_input(input);
  let res1 = part1(&gifts);
  println!("2015 Day 2, Part 1: {res1}");
  let res2 = part2(&gifts);
  println!("2015 Day 2, Part 2: {res2}");
}

#[allow(unused_imports)]
pub mod tests {
  use super::{part1, part2, parse_input};

  #[test]
  fn test_part1() {
    assert_eq!(58, part1(&parse_input("2x3x4")));
    assert_eq!(43, part1(&parse_input("1x1x10")));
  }

  #[test]
  fn test_part2() {
    assert_eq!(34, part2(&parse_input("2x3x4")));
    assert_eq!(14, part2(&parse_input("1x1x10")));
  }
}
